types/src/account_address.rs

```diff
@@ -81,6 +83,52 @@ impl AccountAddress {
     }
 }
 
+/// A struct that represents an account authentication key. An account's address is the last 16
+/// bytes of authentication key used to create it
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Hash, Clone, Copy, CryptoHasher)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+pub struct AuthenticationKey([u8; AUTHENTICATION_KEY_LENGTH]);
+
+impl AuthenticationKey {
+    pub const fn new(bytes: [u8; AUTHENTICATION_KEY_LENGTH]) -> Self {
+        Self(bytes)
+    }
+
+    pub fn random() -> Self {
+        let mut rng = OsRng::new().expect("can't access OsRng");
+        let buf: [u8; AUTHENTICATION_KEY_LENGTH] = rng.gen();
+        AuthenticationKey::new(buf)
+    }
+
+    /// Create an authentication key from a public key by taking its sha3 hash
+    pub fn from_public_key<PublicKey: VerifyingKey>(public_key: &PublicKey) -> Self {
+        Self(*HashValue::from_sha3_256(&public_key.to_bytes()).as_ref())
+    }
+
+    /// Return the last 16 bytes of this authentication key
+    pub fn derived_address(&self) -> AccountAddress {
+        // keep only last 16 bytes
+        let mut array = [0u8; ADDRESS_LENGTH];
+        array.copy_from_slice(&self.0[ADDRESS_LENGTH..AUTHENTICATION_KEY_LENGTH]);
```

Comment:
Maybe use `&self.0[AUTHENTICATION_KEY_LENGTH - ADDRESS_LENGTH..]`
to ensure `ADDRESS_LENGTH` is not used as an index, but for length purposes. Then, even if we ever update any of the two, we are guaranteed to get a suffix of expected length = `ADDRESS_LENGTH`.